tests/Functional/Types/DateTimeTest.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Tests\Functional\Types;
+
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Types\Types;
+
+use function get_class;
+use function sprintf;
+
+final class DateTimeTest extends FunctionalTestCase
+{
+    /** @return string[][] */
+    public static function dataValuesProvider(): array
+    {
+        return [
+            ['2013-04-15 10:10:10.1234567'],
+            ['2013-04-15 10:10:10.123456'],
+            ['2013-04-15 10:10:10.12345'],
+            ['2013-04-15 10:10:10.1234'],
+            ['2013-04-15 10:10:10.123'],
+            ['2013-04-15 10:10:10.12'],
+            ['2013-04-15 10:10:10.1'],
+            ['2013-04-15 10:10:10'],
+
+            ['2013-04-15 10:10:10.9876543'],
+            ['2013-04-15 10:10:10.9876549'],
+            ['2013-04-15 10:10:10.987654'],
+            ['2013-04-15 10:10:10.98765'],
+            ['2013-04-15 10:10:10.9876'],
+            ['2013-04-15 10:10:10.987'],
+            ['2013-04-15 10:10:10.98'],
+            ['2013-04-15 10:10:10.9'],
+
+            ['1999-12-31 23:59:59.99999'],
+        ];
+    }
+
+    /** @dataProvider dataValuesProvider */
+    public function testInsertAndRetrieveDateTime(string $expected): void
+    {
+        $platform = $this->connection->getDatabasePlatform();
+
+        if (! $platform->supportsVariablePrecisionTime()) {
+            self::markTestSkipped(sprintf("Platform %s doesn't support variable precision time", get_class($platform)));
+        }
+
+        $table = new Table('datetime_test_table');
+
+        $vals  = [];
+        $types = [];
+
+        for ($i = 0; $i <= 6; $i++) {
+            $table->addColumn('val' . $i, Types::DATETIME_MUTABLE, ['length' => $i]);
+            $vals['val' . $i]  = $expected;
+            $types['val' . $i] = Types::STRING;
+        }
+
+        $this->dropAndCreateTable($table);
+
+        $this->connection->insert('datetime_test_table', $vals, $types);
+
+        for ($i = 0; $i <= 6; $i++) {
+            $value = Type::getType(Types::DATETIME_MUTABLE)->convertToPHPValue(
+                $this->connection->fetchOne('SELECT val' . $i . ' FROM datetime_test_table'),
+                $platform,
+            );
+
+            $expected = Type::getType(Types::DATETIME_MUTABLE)->convertToPHPValue(
+                $expected,
+                $platform,
+            );
+
+            // PHP stores datatimes with microsecond precision so there will be a difference when column precision is
+            // less than 6. There is a possible one microsecond difference when a database rounds up a time with more
+            // than microsecond precision (PHP truncates it) so allow an additional microsecond difference (i.e. do
+            // not subtract one microsecond from the allowed difference).
```

Comment:
That whole comment block is confusing. You're talking about the database storing dates at a higher precision than PHP (which is 6), yet you don't test with a precision of 7 or higher.

Reply:
Three of the data values are seven decimal places with differing rounding/truncation outcomes so I think I do test precisions higher than six!  The comment was to remind me why I'd not used 10^n/2 - 1 as a threshold as would normally be used in a rounding test. I'll remove all except the first sentence of the comment as it obviously isn't helpful.